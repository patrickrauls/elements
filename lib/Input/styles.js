import { css } from 'styled-components';

const styles = css`
  @keyframes fadein {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }
  position: relative;
  border-bottom: 0.125rem solid #474f4f;
  width: 100%;
  z-index: 1;
  margin-top: 2.1875rem;
  &.invalid,
  &.required {
    color: #af1a12;
    border-bottom: 0.125rem solid #af1a12;
  }
  &.disabled {
    opacity: 0.4;
  }
  .input-requirements {
    color: #021828;
    width: 90%;
    text-align: left;
    animation: fadein 0.3s;
    font-size: 0.875rem;
    font-weight: 200;
    margin: 1rem auto;
    .input-requirement {
      margin: 0.25rem 0;
    }
    .valid-requirement {
      animation: fadein 0.3s;
      margin-left: 0.5rem;
    }
  }
  .input-icon {
    position: absolute;
    height: 1rem;
    width: 0.875rem;
    left: 0;
    bottom: 0.25rem;
  }
  input {
    width: 100%;
    padding: 0 1.125rem 0.25rem;
    border: none;
    border-radius: 0;
    color: #474f4f;
    font-size: 1rem;
  }
  input:focus {
    outline: none;
    color: #021828;
  }
  input:invalid {
    box-shadow: none;
  }
  input:invalid:not(:focus) ~ .flag {
    display: block;
    animation: fadein 0.3s;
  }
  input:focus ~ .flag {
    display: none;
  }
  input::placeholder {
    color: #474f4f;
    opacity: 0.8;
  }
`;

export default styles;
