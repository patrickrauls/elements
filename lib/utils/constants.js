export const FAILED_VALIDATION =
  'Validation requirements have not been satisfied.';

export const regEx = {
  email: /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/
}

export const emails = {
  valid: [
    {
      value: 'email@domain.com',
      description: 'email is standard'
    },
    {
      value: 'firstname.lastname@domain.com',
      description: 'email contains dot in the address field'
    },
    {
      value: 'email@subdomain.domain.com',
      description: 'Email contains dot with subdomain'
    },
    {
      value: 'firstname+lastname@domain.com',
      description: 'Plus sign is considered valid character'
    },
    {
      value: 'email@123.123.123.123',
      description: 'Domain is valid IP address'
    },
    {
      value: 'email@[123.123.123.123]',
      description: 'Square bracket around IP address is considered valid'
    },
    {
      value: '"email"@domain.com',
      description: 'Quotes around email is considered valid'
    },
    {
      value: '1234567890@domain.com',
      description: 'Digits in address are valid'
    }, {
      value: 'email@domain-one.com',
      description: 'Dash in domain name is valid'
    }, {
      value: '_______@domain.com',
      description: 'Underscore in the address field is valid'
    }, {
      value: 'email@domain.name',
      description: '.name is valid Top Level Domain name'
    }, {
      value: 'email@domain.co.uk',
      description: 'Dot in Top Level Domain name also considered valid'
    }, {
      value: 'firstname-lastname@domain.com',
      description: 'Dash in address field is valid'
    }
  ],
  invalid: [
    {
      value: 'plainaddress',
      description: 'Missing @sign and domain'
    }, {
      value: '#@%^%#$@#$@#.com',
      description: 'Garbage'
    }, {
      value: '@domain.com',
      description: 'Missing username'
    }, {
      value: 'email.domain.com',
      description: 'Missing @'
    }, {
      value: 'email.@domain.com',
      description: 'Trailing dot in address is not allowed'
    }, {
      value: 'あいうえお@domain.com',
      description: 'Unicode char as address'
    }, {
      value: 'email@domain',
      description: 'Missing top level domain'
    }, {
      value: 'email@-domain.com',
      description: 'Leading dash in front of domain is invalid'
    }, {
      value: 'email@domain..com',
      description: 'Multiple consecutive dots in the domain portion is invalid'
    }
  ]
};
