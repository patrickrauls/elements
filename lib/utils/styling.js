import { keyframes } from 'styled-components';

export const icons = {
  submit: 'fa fa-chevron-right',
  link: 'fa fa-chevron-right',
  loading: 'fas fa-circle-notch fast-spin',
  big: 'fa-2x',
  valid: 'fa fa-check'
};

export const fastSpin = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(720deg);
  }
`;