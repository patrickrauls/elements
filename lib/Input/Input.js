import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { icons } from '../utils/styling';
import { regEx } from '../utils/constants';
import { FAILED_VALIDATION } from '../utils/constants';
import Flag from '../Flag';
import I from '../I';

class Input extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.defaultValue,
      valid: false,
      dirty: false,
      status: '',
      focus: false
    };
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.validateOn !== this.props.validateOn) {
      this.validate(this.input, nextProps.requirements);
    }
  }
  checkStatus() {
    if (this.state.valid && this.state.dirty) {
      this.setState({ status: 'valid' });
    }
    if (this.props.required && this.state.dirty && !this.state.value) {
      this.setState({ status: 'required' });
    } else if (this.state.dirty && !this.state.valid) {
      this.setState({ status: 'invalid' });
    }
  }
  emailValidation(value) {
    return this.props.type === 'email' ? regEx.email.test(value) : true;
  }
  satisfiesRequirements(value, requirements = this.props.requirements) {
    return requirements
      ? requirements.filter(
        requirement => requirement.validate(value) === false
      ).length === 0
      : true;
  }
  validate(input, requirements) {
    if (
      this.satisfiesRequirements(input.value, requirements) &&
      this.emailValidation(input.value)
    ) {
      input.setCustomValidity('');
    } else {
      input.setCustomValidity(FAILED_VALIDATION);
    }
    this.setState(
      {
        valid: input.checkValidity(),
        value: input.value
      },
      () => {
        this.checkStatus();
      }
    );
  }
  handleChange = e => {
    if (this.props.onChange) {
      this.props.onChange(e);
    }
    this.validate(e.target, this.props.requirements);
    this.setState({ dirty: true });
  };
  handleClick = ({ currentTarget }) => {
    const input = [...currentTarget.children].find(element => {
      return element.tagName === 'INPUT';
    })
    input.focus();
  };
  handleFocus = e => {
    if (this.props.onFocus) {
      this.props.onFocus(e);
    }
    this.setState({
      focus: true,
      dirty: true
    });
  };
  handleBlur = e => {
    if (this.props.onBlur) {
      this.props.onBlur(e);
    }
    this.setState(
      {
        focus: false
      },
      () => {
        this.checkStatus();
      }
    );
  };
  render() {
    const {
      className,
      disabled,
      icon,
      inputMode,
      name,
      required,
      requirements,
      spellCheck,
      type,
      // destructure the following out of the ...rest object.
      defaultValue, validateOn, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    return (
      <div
        className={`
          ${className}
          ellies-input-container
          ${this.state.focus ? 'ellies-input-focused' : ''}
          ${type}-input-component
          ${this.state.status}
          ${disabled ? 'disabled' : ''}
        `}
        onClick={this.handleClick}
        onKeyDown={this.handleClick}
        role="presentation"
      >
        {requirements &&
          this.state.focus && (
          <ul className="ellies-input-requirements">
            {requirements.map(requirement => (
              <li className="ellies-input-requirement" key={requirement.description}>
                <span className="ellies-input-requirement-description">{requirement.description}</span>
                <I
                  className="ellies-input-requirement-status-icon"
                  hidden={!requirement.validate(this.state.value)}
                  icon={icons.valid}
                />
              </li>
            ))}
          </ul>
        )}
        {icon && (
          <I
            className="ellies-input-icon"
            icon={icon}
          />
        )}
        <input
          {...rest}
          className={`${name}-${type}-input input-element`}
          disabled={disabled}
          inputMode={type === 'email' ? 'email' : inputMode}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          ref={input => {
            this.input = input;
          }}
          required={required}
          spellCheck={type === 'email' ? false : spellCheck}
          type={type === 'email' ? 'text' : type}
          value={this.state.value}
        />
        <Flag type={this.state.status || 'hidden'} />
      </div>
    );
  }
}

Input.propTypes = {
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  inputMode: PropTypes.string,
  name: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  required: PropTypes.bool,
  requirements: PropTypes.arrayOf(
    PropTypes.shape({
      description: PropTypes.string.isRequired,
      validate: PropTypes.func.isRequired
    })
  ),
  spellCheck: PropTypes.bool,
  type: PropTypes.oneOf([
    'email',
    'number',
    'password',
    'search',
    'tel',
    'text',
    'url'
  ]),
  validateOn: PropTypes.any
};

Input.defaultProps = {
  className: '',
  onFocus: undefined,
  name: undefined,
  type: 'text',
  defaultValue: '',
  disabled: false,
  icon: '',
  inputMode: undefined,
  spellCheck: undefined,
  requirements: undefined,
  validateOn: undefined,
  required: false,
  onBlur: undefined,
  onChange: undefined
};

export default Input;
