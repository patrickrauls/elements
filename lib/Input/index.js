import styled from 'styled-components';
import Input from './Input';
import styles from './styles';

const StyledInput = styled(Input)`
  ${styles};
`;

export default StyledInput;
