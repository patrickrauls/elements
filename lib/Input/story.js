import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';
import Input from './index';
// import withTheme from '../../../stories/withTheme';

storiesOf('Input', module)
  // .addDecorator(withTheme);
  .addDecorator(withKnobs)
  .add('Username', () => (
    <Input
      name="username"
      placeholder="username"
      type={select('type', ['text', 'email', 'number', 'password'])}
      disabled={boolean('disabled', false)}
      onChange={action('changed')}
      required={boolean('required', false)}
    />
  ))
  .add('Password', () => (
    <Input
      name="password"
      placeholder="password"
      type={select('type', ['text', 'email', 'number', 'password'])}
      disabled={boolean('disabled', false)}
      onChange={action('changed')}
      required={boolean('required', false)}
    />
  ));
