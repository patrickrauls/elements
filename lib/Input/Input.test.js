import React from 'react';
import { mount, shallow } from 'enzyme';

import Input from './Input';
import { FAILED_VALIDATION, emails } from '../utils/constants';

describe('Input', () => {
  const setup = props => shallow(<Input {...props} />);
  describe('unit tests', () => {
    let wrapper = setup();
    it('should render without crashing', () => {
      expect(wrapper.exists()).toBe(true);
    });
    it('should make the defaultValue the initial value', () => {
      wrapper = setup({ defaultValue: 'Zebra' });
      expect(wrapper.find('input').prop('value')).toEqual('Zebra');
    });
    it('should set a class of disabled if given a disabled prop', () => {
      wrapper = setup({ disabled: true });
      expect(wrapper.hasClass('disabled')).toBe(true);
    });
    it('should fire a given onBlur function', () => {
      const onBlur = jest.fn();
      wrapper = setup({ onBlur });
      wrapper.find('input').simulate('blur');
      expect(onBlur).toHaveBeenCalled();
    });
    it('should fire a given onFocus function', () => {
      const onFocus = jest.fn();
      wrapper = setup({ onFocus });
      wrapper.find('input').simulate('focus');
      expect(onFocus).toHaveBeenCalled();
    });
    it('should focus on <input /> when container is clicked', () => {
      const focus = jest.fn();
      const currentTarget = {
        children: [
          {
            tagName: 'INPUT',
            focus
          }
        ]
      };
      wrapper = setup();
      wrapper.simulate('click', { currentTarget });
      expect(focus).toHaveBeenCalled();
    });
    it('should display each requirement description on focus', () => {
      const description = 'must be Jamaica';
      const validate = value => value === 'Jamaica';
      const requirements = [{ description, validate }];
      wrapper = setup({ requirements });
      wrapper.find('input').simulate('focus');
      expect(wrapper.find('.input-requirement').text()).toEqual(description);
    });
    it('should hide each requirement description on blur', () => {
      wrapper.find('input').simulate('blur');
      expect(wrapper.find('.input-requirement').length).toEqual(0);
    });
    it('should indicate when a requirement has been met on change', () => {
      const description = 'must be Jamaica';
      const validate = value => value === 'Jamaica';
      const requirements = [{ description, validate }];
      wrapper = mount(<Input requirements={requirements} />);
      wrapper.find('input').simulate('focus');
      wrapper.find('input').simulate('change', {
        target: {
          value: 'Jamaica',
          setCustomValidity() { },
          checkValidity() { }
        }
      });
      expect(wrapper.find('.input-requirement')
        .find('.valid-requirement').length).toEqual(1);
    });
  });
  describe('interaction tests', () => {
    let wrapper;
    const setCustomValidity = jest.fn();
    const checkValidity = jest.fn();
    const target = {
      value: 'Tiger',
      setCustomValidity,
      checkValidity
    };
    it('should fire a given onChange function', () => {
      const onChange = jest.fn();
      wrapper = setup({ onChange });
      wrapper.find('input').simulate('change', { target });
      expect(onChange).toHaveBeenCalled();
    });
    it('should sync value on change', () => {
      wrapper = setup({ defaultValue: 'Zebra' });
      wrapper.find('input').simulate('change', { target });
      expect(wrapper.find('input').prop('value')).toEqual('Tiger');
    });
    it('should fire validations from requirements prop on change', () => {
      const validate = jest.fn();
      const anotherValidate = jest.fn();
      wrapper = setup({
        requirements: [{
          description: 'mock requirement',
          validate
        }, {
          description: 'another mock requirement',
          validate: anotherValidate
        }]
      });
      wrapper.find('input').simulate('change', { target });
      expect(validate).toHaveBeenCalledWith(target.value);
      expect(anotherValidate).toHaveBeenCalledWith(target.value);
    });
    it('should set itself as invalid on change for unmet validations', () => {
      const validate = value => false;
      wrapper = setup({
        requirements: [{
          description: 'mock requirement',
          validate
        }]
      });
      wrapper.find('input').simulate('change', { target });
      expect(setCustomValidity).toHaveBeenCalledWith(FAILED_VALIDATION);
    });
    it('unsets itself as invalid on change when validations are met', () => {
      const validate = value => true;
      wrapper = setup({
        requirements: [{
          description: 'mock requirement',
          validate
        }]
      });
      wrapper.find('input').simulate('change', { target });
      expect(setCustomValidity).toHaveBeenCalledWith('');
    });
  });
  describe('UI tests', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should re-validate any time the validateOn prop changes', () => {
      const validate = jest.fn();
      const props = {
        validateOn: 'Foo',
        defaultValue: 'Zoo',
        requirements: [{
          description: 'mock validation',
          validate
        }]
      };
      wrapper = mount(<Input {...props} />);
      wrapper.setProps({ validateOn: 'Bar' });
      expect(validate).toHaveBeenCalledWith('Zoo');
    });
    it('should not validate without a change in value or validateOn', () => {
      const validate = jest.fn();
      const props = {
        validateOn: 'Foo',
        defaultValue: 'Zoo',
        requirements: [{
          description: 'mock validation',
          validate
        }]
      };
      wrapper = mount(<Input {...props} />);
      wrapper.setProps({ validateOn: 'Foo' });
      expect(validate).not.toHaveBeenCalled();
    });
    it('should provide a prop type=valid to a Flag if valid', () => {
      wrapper = mount(<Input />);
      wrapper.find('input').simulate('change');
      expect(wrapper.find('Flag').prop('type')).toEqual('valid');
    });
    it('should provide a prop type=invalid to a Flag if invalid', () => {
      wrapper = mount(
        <Input
          requirements={[{
            description: 'fails to validate',
            validate: value => false
          }]}
        />
      );
      wrapper.find('input').simulate('change');
      expect(wrapper.find('Flag').prop('type')).toEqual('invalid');
    });
    it('should provide a prop type=required to a Flag if required', () => {
      wrapper = mount(<Input required />);
      wrapper.find('input').simulate('change');
      expect(wrapper.find('Flag').prop('type')).toEqual('required');
    });
  });
  describe('email input validation tests', () => {
    describe('valid emails', () => {
      const wrapper = mount(<Input type="email" />);
      afterAll(() => {
        wrapper.unmount();
      });
      const input = wrapper.find('input');
      const changeTo = (value) => {
        input.instance().value = value;
        input.simulate('change');
      };
      emails.valid.forEach(email => {
        it(email.description, () => {
          changeTo(email.value);
          expect(wrapper.find('Flag').prop('type')).toEqual('valid');
        });
      });
    });
    describe('invalid emails', () => {
      const wrapper = mount(<Input type="email" />);
      afterAll(() => {
        wrapper.unmount();
      });
      const input = wrapper.find('input');
      const changeTo = (value) => {
        input.instance().value = value;
        input.simulate('change');
      };
      emails.invalid.forEach(email => {
        it(email.description, () => {
          changeTo(email.value);
          expect(wrapper.find('Flag').prop('type')).toEqual('invalid');
        });
      });
    })
  });
});
/*
*/
