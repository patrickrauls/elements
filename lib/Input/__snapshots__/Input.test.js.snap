// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input ui tests should add styles to the Input component 1`] = `
<Input
  className="sc-bwzfXH crTysL"
  defaultValue=""
  disabled={false}
  required={false}
  type="text"
/>
`;
